Ensure decrypted internal keys never swapped

dAdAbird opened the pull request:
We have already had a locked memory for internal keys. But we read and created keys into usually palloced memory before copying them to the locked memory.

This commit changes the approach. Now, the user must allocate (acquire) a locked memory first and use it as a buffer for decryption/creating the key. This locked memory is kinda generic, so the user requests an amount of bytes (rather than objects). Although currently, during the key search, we assume that everything the memory contains only `RelKeyCacheRec` objects. With two exceptions: 1) server creates its own locked page to store the WAL write key;
2) `pg_tde_perform_rotate_key` cheekily allocates a space there for the keys re-encryption but releases this memory when it's done.

In future, if we switch to the full _map files encryption, we can gulp (mmap) the whole file into that memory.

Fixes PG-1445

dAdAbird committed:
Ensure decrypted internal keys never swapped

We have already had a locked memory for internal keys. But we read and
created keys into usually palloced memory before copying them to the
locked memory.

This commit changes the approach. Now, the user must allocate (acquire)
a locked memory first and use it as a buffer for decryption/creating
the key. This locked memory is kinda generic, so the user requests an
amount of bytes (rather than objects). Although currently, during the
key search, we assume that everything the memory contains only
`RelKeyCacheRec` objects. With two exceptions: 1) server creates its
own locked page to store the WAL write key;
2) `pg_tde_perform_rotate_key` cheekily allocates a space there for
the keys re-encryption but releases this memory when it's done.

In future, if we switch to the full _map files encryption, we can gulp
(mmap) the whole file into that memory.

Fixes PG-1445

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_encrypt.c:
```diff
@@ -163,48 +158,72 @@ TDEXLogWriteEncryptedPages(int fd, const void *buf, size_t count, off_t offset,
+#else
+		pageSize = sysconf(_SC_PAGESIZE);
+#endif
+		EncryptionKey = (InternalKey *) MemoryContextAllocAligned(TopMemoryContext, pageSize, pageSize, MCXT_ALLOC_ZERO);
```
Isn't it cleaner to just call `MemoryContextAllocExtended()` directly?

dAdAbird replied:
We need the pointer to be aligned to the page pageSize. `MemoryContextAllocExtended` can't do that

jeltz replied:
Sorry, misread the code.

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_encrypt.c:
```diff
@@ -163,48 +158,72 @@ TDEXLogWriteEncryptedPages(int fd, const void *buf, size_t count, off_t offset,
+	 */
+	if (key)
+	{
+		long		pageSize = 0;
```
There is no reason to have a default value if we are always assigning it a couple of lines below.

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_encrypt.c:
```diff
@@ -163,48 +158,72 @@ TDEXLogWriteEncryptedPages(int fd, const void *buf, size_t count, off_t offset,
+		pageSize = getpagesize();
+#else
+		pageSize = sysconf(_SC_PAGESIZE);
+#endif
```
PostgreSQL itself always calls `sysconf(_SC_PAGESIZE);`  so it is apparently guaranteed to be supported on all platofrms except Windows. Do we need the fallback to `getpagesize();`?

dAdAbird replied:
Yes, you are right, we should drop getpagesize(). I was playing safe but it is a legacy call and was removed in POSIX.1-2001. Doesn't make sense to have it

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_encrypt.c:
```diff
@@ -222,15 +241,15 @@ tdeheap_xlog_seg_write(int fd, const void *buf, size_t count, off_t offset,
 	 * This func called with WALWriteLock held, so no need in any extra sync.
 	 */
-	if (EncryptionKey.rel_type & TDE_KEY_TYPE_GLOBAL &&
+	if (EncryptionKey && EncryptionKey->rel_type & TDE_KEY_TYPE_GLOBAL &&
```
With this change can't we remove the check for `rel_type & TDE_KEY_TYPE_GLOBAL` since all keys will have that flag. The reaosn we needed it was to handle if `EncryptionKey` had not been set.

jeltz commented on contrib/pg_tde/t/009_wal_encrypt.pl:
```diff
@@ -55,10 +55,28 @@
+    DECLARE idx integer;
+begin
+    for idx in 0..700 loop
+        EXECUTE format('CREATE TABLE t%s (c1 int) USING tde_heap', idx);
```
Not that it matters in tests but in real production code I would have done.

```
EXECUTE format('CREATE TABLE %I (c1 int) USING tde_heap', 't' || idx);
```

dAdAbird replied:
So should I change it? I took it from the `cache_alloc` regression test. If to change, that it make sense to fix the regression test as well

jeltz replied:
Up to you. I just wanted to share how real production code looks like.

jeltz commented on contrib/pg_tde/src/access/pg_tde_tdemap.c:
```diff
@@ -76,50 +76,52 @@ typedef struct RelKeyCacheRec
 static WALKeyCacheRec *pg_tde_add_wal_key_to_cache(InternalKey *cached_key, XLogRecPtr start_lsn);
-static InternalKey *pg_tde_put_key_into_cache(const RelFileLocator *locator, InternalKey *key);
+static void *tde_alloc_safe_mem(size_t len);
+static inline void tde_release_safe_mem(size_t len);
```
I am not the biggest fan of explicit inline unless there is a good reason for it. I think the optimizer would inline this when it should anyway.

jeltz commented on contrib/pg_tde/src/access/pg_tde_tdemap.c:
```diff
@@ -76,50 +76,52 @@ typedef struct RelKeyCacheRec
+static void *tde_alloc_safe_mem(size_t len);
+static inline void tde_release_safe_mem(size_t len);
+
+#define AllocRelCache(SIZE) (RelKeyCacheRec *) tde_alloc_safe_mem(sizeof(RelKeyCacheRec) * SIZE);
```
Since we always allocate 1 is there any reason for this macro? At least in its current form. Maybe it should be like `palloc_object()` instead?

dAdAbird replied:
Not always but most of the time. In pg_tde_fetch_wal_keys we allocate all the keys at once. But I'm not sure about this macros either. I propose to leave it like that for now since there are might to be changes around that due to "Principla key and Keyring protection" (PG-1528) anyway. And I'll revisit it there

jeltz commented on contrib/pg_tde/src/access/pg_tde_tdemap.c:
```diff
@@ -149,11 +151,12 @@ pg_tde_create_smgr_key(const RelFileLocatorBackend *newrlocator)
 pg_tde_create_key_map_entry(const RelFileLocator *newrlocator, uint32 entry_type)
 {
-	InternalKey rel_key_data;
+
```
You accidentally added an empty line.

jeltz commented on contrib/pg_tde/src/access/pg_tde_tdemap.c:
```diff
@@ -209,38 +215,42 @@ tde_sprint_key(InternalKey *k)
 
 /*
- * Generates a new internal key for WAL and adds it to the _dat file. It doesn't
+ * Generates a new internal key for WAL and adds it to the _map file. It doesn't
```
This should be fixed in a separate commit.

jeltz commented on contrib/pg_tde/src/access/pg_tde_tdemap.c:
```diff
@@ -209,38 +215,42 @@ tde_sprint_key(InternalKey *k)
 	{
 		ereport(ERROR,
-				(errmsg("principal key not configured"),
+				(errmsg("principal key is not configured"),
```
If this should be fixed it should be fixed in all places and as a separate commit.

jeltz commented on contrib/pg_tde/src/access/pg_tde_tdemap.c:
```diff
@@ -698,16 +708,16 @@ pg_tde_perform_rotate_key(TDEPrincipalKey *principal_key, TDEPrincipalKey *new_p
-		pfree(rel_key_data);
 	}
 
+	tde_release_safe_mem(sizeof(InternalKey));
```
This API is quite dangerous, maybe we should actually pass the pointer to assert that we actually release from the end.

jeltz commented on contrib/pg_tde/src/access/pg_tde_tdemap.c:
```diff
@@ -1460,16 +1470,20 @@ pg_tde_add_wal_key_to_cache(InternalKey *cached_key, XLogRecPtr start_lsn)
 	static long pageSize = 0;
-	RelKeyCacheRec *rec;
 	MemoryContext oldCtx;
+	int			needPages = 0;
```
No reason to have a default here.

jeltz commented on contrib/pg_tde/src/access/pg_tde_tdemap.c:
```diff
@@ -1460,16 +1470,20 @@ pg_tde_add_wal_key_to_cache(InternalKey *cached_key, XLogRecPtr start_lsn)
-	RelKeyCacheRec *rec;
 	MemoryContext oldCtx;
+	int			needPages = 0;
+	unsigned char *newPages = NULL;
```
Same hjere.

jeltz commented on contrib/pg_tde/src/access/pg_tde_tdemap.c:
```diff
@@ -1480,77 +1494,55 @@ pg_tde_put_key_into_cache(const RelFileLocator *rlocator, InternalKey *key)
-
-		tde_rel_key_cache.len = 0;
-		tde_rel_key_cache.cap = (pageSize - 1) / sizeof(RelKeyCacheRec);
+		goto RETURN;
```
PostgreSQL uses lowercase for labels.

jeltz reviewed (changes requested):
I am not the biggest fan of this approach of heavy interaction between the cache and storage. Especially if we want to change from linear cache to hash map for example. But it is not like I have a better suggestion which does not involve writing our own allocator.

Plus I think the keys in `TDESMgrRelationData` still can be swapped to disk.

dAdAbird committed:
Fetch smgr key from TDE cache instead of TDESMgrRelation

Smgr has its own cache, TDESMgrRelation, which stores the state and other. We use it to store the Internal key. Although we have to store a pointer to the key in the secure memory rather than a copy of the key. But when secure memory grows, it moves the object, therefore, the smgr pointer becomes invalid. Moreover, pointers become invalid even w/o growth. The simplest scenario:
```
CREATE TEMP TABLE articles (
    id int CONSTRAINT articles_pkey PRIMARY KEY,
    keywords text,
    title text UNIQUE NOT NULL,
    body text UNIQUE,
    created date
);

SELECT id, keywords, title, body, created FROM articles GROUP BY id;
ERROR:  invalid page in block 0 of relation base/16384/t0_32940
```

This commit removes the internal key pointer from smgr and replaces it with a GetSMGRRelationKey call. By the time of this call, the key would be in a cache. Though this will have performance drawbacks and should be benchmarked and re-evaluated

jeltz commented:
I think we can close this for now. We learned a lot useful from this.
